posts/extend-dataframes/index.qmd

```diff
@@ -0,0 +1,283 @@
+
+```{r}
+validate_birthdays <- function(x) {
+  stopifnot(
+    "input must contain 'name' and 'birthday' columns" =
+      all(c("name", "birthday") %in% colnames(x)),
+    "names must be a character" =
+      is.character(x$name),
+    "birthday must be a date" =
+      lubridate::is.Date(x$birthday)
+  )
+  invisible(x)
+}
+```
+
+
+This will return an error if the class is not valid (defined in terms of the class' invariants). 
+
+Now we can show how to error if one of the invariants are removed during subsetting. 
+
+```{r, eval=FALSE}
+`[.birthdays` <- function(x) {
+  validate_birthdays(NextMethod())
+}
+
+birthdays[, -1]
+#  Error in validate_birthdays(NextMethod()) :
+#  input must contain 'name' and 'birthday' columns
+```
+
+The second design option is the reconstruct-on-subsetting. This checks whether the class is valid, and if not downgrade the class to the superclass, in our case a data frame. This is done by not only validating the object during subsetting but to check whether it is a valid class object, and then either ensuring all of the attributes of the subclass -- in our case `<birthdays>` -- are maintained, or attributes are stripped and only the attributes of the base superclass -- in our case `data.frame` -- are kept. 
+
+
+
+::: {.callout-note}
+Important note: this section of the post relies heavily on [https://github.com/DavisVaughan/2020-06-01_dplyr-vctrs-compat](https://github.com/DavisVaughan/2020-06-01_dplyr-vctrs-compat).
+:::
+  
+  The four functions that are required to be added to ensure our class is correctly handled when invaliding it are: 
+  
+  - `birthdays_reconstruct()`
+- `birthdays_can_reconstruct()`
+- `df_reconstruct()`
+- `dplyr_reconstruct.birthdays()`
+
+We'll tackle the first three first, and then move onto to the last one as this requires some extra steps. 
+
+`birthdays_reconstruct()` is a function that contains an if-else statement to determine whether the returned object is a `<birthdays>` or `data.frame` object.
+
+```{r}
+birthdays_reconstruct <- function(x, to) {
+  if (birthdays_can_reconstruct(x)) {
+    df_reconstruct(x, to)
+  } else {
+     x <- as.data.frame(x)
+    message("Removing crucial column in `<birthdays>` returning `<data.frame>`")
+    x
+  }
+}
+```
+
+The if-else evaluation is controlled by `birthdays_can_reconstruct()`. This function determines whether after subsetting the object is a valid `<birthdays>` class. It checks whether the validator fails, in which case it returns `FALSE`, otherwise the function will return `TRUE`.
+
+```{r}
+birthdays_can_reconstruct <- function(x) {
+  # check whether input is valid
+  valid <- tryCatch(
+    { validate_birthdays(x) },
+    error = function(cnd) FALSE
+  )
+
+  # return boolean
+  !isFALSE(valid)
```

Comment:
>  !isFALSE(valid)

This seems a bit convoluted to me. Why are you not returning `valid` as is? If the `birthdays_can_reconstruct()` is "validating" the statement, "it is possible to reconstruct `birthday`", then the result of `tryCatch()`, in my opinion, should answer the question and not undergo further checks and negations.

Reply:
Thanks for raising as the logical inverting can be tricky to read. The `!isFALSE()` syntax is less to control for the `tryCatch()` returning `FALSE` and more for when it returns a `birthdays` object invisibly. In which cases the `!isFALSE()` will return TRUE.

Please provide a code chunk of possible alternative formulations that I could replace the code with.

Reply:
Oh, I see. That makes sense. I'm happy for it to be left as is.